File: backend/app/schemas/features.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Feature row (single date)
# ---------------------------------------------------------------------------

class FeatureRow(BaseModel):
    date: str
    close: float | None = None
    # Trend
    sma_10:  float | None = None
    sma_20:  float | None = None
    sma_50:  float | None = None
    sma_200: float | None = None
    ema_10:  float | None = None
    ema_20:  float | None = None
    ema_50:  float | None = None
    macd_line:   float | None = None
    macd_signal: float | None = None
    macd_hist:   float | None = None
    adx_14:      float | None = None
    price_vs_sma20_pct: float | None = None
    price_vs_sma50_pct: float | None = None
    # Momentum
    rsi_14:    float | None = None
    stoch_k:   float | None = None
    stoch_d:   float | None = None
    roc_10:    float | None = None
    williams_r: float | None = None
    # Volatility
    bb_upper:    float | None = None
    bb_lower:    float | None = None
    bb_width:    float | None = None
    bb_pct_b:    float | None = None
    atr_14:      float | None = None
    hist_vol_20: float | None = None
    # Volume
    obv:        int   | None = None
    vol_sma_20: float | None = None
    vol_ratio:  float | None = None
    cmf_20:     float | None = None
    # Price action
    daily_return:    float | None = None
    return_3d:       float | None = None
    return_5d:       float | None = None
    return_10d:      float | None = None
    gap_pct:         float | None = None
    hl_range_pct:    float | None = None
    candle_body_pct: float | None = None

    model_config = {"from_attributes": True}


class FeatureListResponse(BaseModel):
    ticker: str
    count:  int
    rows:   list[FeatureRow]


# ---------------------------------------------------------------------------
# Feature statistics
# ---------------------------------------------------------------------------

class FeatureStat(BaseModel):
    mean:     float | None
    std:      float | None
    min:      float | None
    max:      float | None
    null_pct: float | None


class FeatureStatsResponse(BaseModel):
    ticker:     str
    total_rows: int
    stats:      dict[str, FeatureStat]


# ---------------------------------------------------------------------------
# Labels
# ---------------------------------------------------------------------------

class LabelDistribution(BaseModel):
    ticker: str
    total:  int
    BUY:    int
    SELL:   int
    HOLD:   int


# ---------------------------------------------------------------------------
# Dataset versions
# ---------------------------------------------------------------------------

class DatasetVersionInfo(BaseModel):
    version:          int
    tickers:          list[str]
    date_range_start: str | None
    date_range_end:   str | None
    n_rows:           int | None
    n_train:          int | None
    n_test:           int | None
    feature_list:     list[str]
    label_distribution: dict[str, int]
    split_date:       str | None
    file_path:        str | None
    created_at:       str | None

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Pipeline
# ---------------------------------------------------------------------------

class PipelineRunRequest(BaseModel):
    tickers:     list[str] | None = None
    split_date:  str | None       = None
    build_dataset: bool           = True


class PipelineRunResponse(BaseModel):
    message:            str
    tickers_processed:  list[str]
    tickers_skipped:    list[str]
    dataset_version:    int | None
    label_distribution: dict[str, int] | None

File: frontend/lib/features.ts
export const FEATURE_GROUPS = {
  trend: [
    "sma_10", "sma_20", "sma_50", "sma_200",
    "ema_10", "ema_20", "ema_50",
    "macd_line", "macd_signal", "macd_hist", "adx_14",
    "price_vs_sma20_pct", "price_vs_sma50_pct",
  ],
  momentum:  ["rsi_14", "stoch_k", "stoch_d", "roc_10", "williams_r"],
  volatility: ["bb_upper", "bb_lower", "bb_width", "bb_pct_b", "atr_14", "hist_vol_20"],
  volume:    ["obv", "vol_sma_20", "vol_ratio", "cmf_20"],
  price_action: [
    "daily_return", "return_3d", "return_5d", "return_10d",
    "gap_pct", "hl_range_pct", "candle_body_pct",
  ],
} as const;

export type FeatureGroup = keyof typeof FEATURE_GROUPS;

export const ALL_FEATURES: string[] = Object.values(FEATURE_GROUPS).flat();

export const GROUP_LABELS: Record<FeatureGroup, string> = {
  trend:        "Trend",
  momentum:     "Momentum",
  volatility:   "Volatility",
  volume:       "Volume",
  price_action: "Price Action",
};

export const GROUP_COLORS: Record<FeatureGroup, string> = {
  trend:        "#f97316",   // orange
  momentum:     "#a855f7",   // purple
  volatility:   "#64748b",   // slate
  volume:       "#3b82f6",   // blue
  price_action: "#10b981",   // emerald
};

/** Human-readable label for a feature key */
export function featureLabel(key: string): string {
  return key
    .replace(/_/g, " ")
    .replace(/\b\w/g, (c) => c.toUpperCase())
    .replace(/Sma/g, "SMA")
    .replace(/Ema/g, "EMA")
    .replace(/Macd/g, "MACD")
    .replace(/Rsi/g, "RSI")
    .replace(/Adx/g, "ADX")
    .replace(/Atr/g, "ATR")
    .replace(/Bb/g, "BB")
    .replace(/Obv/g, "OBV")
    .replace(/Cmf/g, "CMF")
    .replace(/Pct/g, "%")
    .replace(/Vs/g, "vs");
}
